codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/transformers/RefactorConstrainedMemberType.kt

```diff
@@ -0,0 +1,195 @@
+                walker.walkShapes(it)
+            }
+            .filterIsInstance(StructureShape::class.java)
+            .flatMap {
+                it.constrainedMembers()
+            }.mapNotNull { it.makeNonConstrained(model, existingRefactoredNames) }
+
+        return applyTransformations(model, transformations)
+    }
+
+    /***
+     * Returns a Model that has all the transformations applied on the original
+     * model.
+     */
+    private fun applyTransformations(
+        model: Model,
+        transformations: List<MemberTransformation>,
+    ): Model {
+        if (transformations.isEmpty())
+            return model
+
+        val modelBuilder = model.toBuilder()
+        val memberShapesToChange: MutableList<MemberShape> = mutableListOf()
+
+        transformations.forEach {
+            modelBuilder.addShape(it.newShape)
+
+            val changedMember = it.memberToChange.toBuilder()
+                .target(it.newShape.id)
+                .traits(it.traitsToKeep)
+                .build()
+            memberShapesToChange.add(changedMember)
+        }
+
+        // Change all original constrained member shapes with the new standalone types,
+        // and keep only the non-constraint traits on the member shape.
+        return ModelTransformer.create()
+            .replaceShapes(modelBuilder.build(), memberShapesToChange)
+    }
+
+    /**
+     * Returns a list of members that have constraint traits applied to them
+     */
+    private fun StructureShape.constrainedMembers(): List<MemberShape> =
+        this.allMembers.values.filter {
+            it.hasConstraintTrait()
+        }
+
+    /**
+     * Returns the unique (within the model) name of the refactored shape
+     */
+    private fun refactoredShapeId(
+        model: Model,
+        existingRefactoredNames: MutableSet<ShapeId>,
+        memberShape: ShapeId,
+    ): ShapeId {
+        val structName = memberShape.name
+        val memberName = memberShape.member.orElse(null)
+            .replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }
+
+        fun makeStructName(suffix: String = "") =
+            ShapeId.from("${memberShape.namespace}#Refactored${structName}${memberName}$suffix")
+
+        fun structNameIsUnique(newName: ShapeId) =
+            model.getShape(newName).isEmpty && !existingRefactoredNames.contains(newName)
+
+        fun generateUniqueName(): ShapeId {
+            // Ensure the name does not already exist in the model, else make it unique
+            // by appending a new number as the suffix.
+            var suffix = ""
+            (0..100).forEach {
+                val extractedStructName = makeStructName(suffix)
+                if (structNameIsUnique(extractedStructName))
+                    return extractedStructName
+
+                suffix = "_$it"
+            }
+
+            // A unique type could not be found by adding a numeric suffix to it. As a
```

Comment:
I think I'd abort here; the main algorithm should be good enough and not cause collisions.